File: src/ports/C++/interfaces/fltk/mark_cell.cxx
#include <FL/Fl.H>
#include <FL/fl_draw.H>

#include "fltk/mark_cell.hh"

GameOfLife::FLTK::MarkCell::MarkCell(long int width, long int height) : WIDTH(width), HEIGHT(height), SIZE(4), ALIVE(fl_rgb_color(220, 80, 120)), DEAD(fl_rgb_color(210, 240, 190))
{
}

void GameOfLife::FLTK::MarkCell::aliveAt(long int x, long int y)
{
	markCellAt(x, y, ALIVE);
}

void GameOfLife::FLTK::MarkCell::deadAt(long int x, long int y)
{
	markCellAt(x, y, DEAD);
}

void GameOfLife::FLTK::MarkCell::markCellAt(long int x, long int y, Fl_Color colour)
{
  long int x1 = x * SIZE + WIDTH / 2 - SIZE / 2;
  long int y1 = y * SIZE + HEIGHT / 2 - SIZE / 2;

  fl_rectf(x1, y1, SIZE - 1, SIZE - 1, colour);
}

File: src/impl/ruby/lib/interfaces/fox/mark_cell.rb
require 'fox16'

module GameOfLife
  module Interfaces
    module Fox
      class MarkCell
        SIZE = 4
        ALIVE_COLOUR = ::Fox::FXRGB(204, 51, 102)
        DEAD_COLOUR = ::Fox::FXRGB(172, 200, 172)

        attr_reader :context, :width, :height

        def initialize options
          @context = options[:context]
          @width = options[:width]
          @height = options[:height]
        end

        def alive_at x, y, generation
          draw_at x, y, ALIVE_COLOUR
        end

        def dead_at x, y, generation
          draw_at x, y, DEAD_COLOUR
        end

        def draw_at x, y, colour
          x1 = x * SIZE + width / 2 - SIZE / 2
          y1 = y * SIZE + height / 2 - SIZE / 2

          context.setForeground colour
          context.fillRectangle x1, y1, SIZE - 1, SIZE - 1
        end
      end
    end
  end
end
